Design/Workbench/design-projectsaving.md

```diff
@@ -0,0 +1,229 @@
+- Be able to save and load, plots, interfaces, workspaces, workspaces interfaces and scripts.
+  - Scripts only lightly and not transferrable across machines/operating systems
+
+Changes to MantidWorkbench needed
+---------------------------------
+- MainWindow should track all of the open windows (as well as if they are serializable or not) similar to MantidPlot
+- File->Save Project needs to be added
+- File->Load Project needs to be added
+- A GUI for picking where to Save
+
+GUI and integration initial ideas
+---------------------------------
+It is true that this document is mostly focused on the actual implementation of saving the project and they intricacies of how that would work. However, it is important that this is discussed to some extent before hand. The intial thought process is that you would be able to do both a basic SaveAll and Save Advanced. SaveAll will run through a pretty simple Save everything as long as it's not overwriting workspaces. Save Advanced would allow the users to Select workspaces and windows individually, also adding an option as to whether or not they want to overwrite the workspaces that may or may be saved which is so at worst case users can save everything from scratch.
+
+Options for Implementation of Workspace Saving
+----------------------------------------------
+**Python:**
+- Pickling
+	- Pros: Built in library of Python
+	- Cons: Need to add _ _ getstate_ _ and _ _ setstate_ _ functions to every UI class which must return a pickle-able type (dict, list, int, string etc). Then setstate function. Unconfirmed if possible or not.
+- JSON Encoding
+	- Pros: Built in library of Python, "faster" than Pickling: https://www.benfrederickson.com/dont-pickle-your-data/, Potential to be integrated with C++ as it's universal
+	- Cons: Needs methods in every class, doesn't integrate fully with C++ (It won't translate a C++ class into a JSON file/string without an extra library)
+
+**C++**
+- Serialization with boost 
+	- Pros: We already have it, simple, uses an archive type for template serialization, has forward compatibility when classes change.
+	- Cons: We need to give Friend access to each UI class for the boost::serialize::access, not python.
+- Serialization with cereal
+	- Pros: Very easy to implement using a template of Archive type.
+	- Cons: We don't already have the cereal library, plus we'd need to cite it, as per licensing, not python.
+
+**Both C++ and Python**
+- Our own method of serialization utilizing a JSON back-end
+	- Pros: We can pass C++ values to Python based serializer and back to C++, we can determine file format, we can write custom conversions for types (Most Qt types have a QVariant type conversion to JSON anyway), highly flexible
+	- Cons: Requires extra work to implement that a already available library 
+
+
+Implementation of save window, the recommended Method
+-----------------------------------------------------
```

Comment:
I'm not so sure about this section - it's not entirely clear how all of the bits fit together. I think a few key points as bullets would help first like:
* an intermediary format will be used for windows to save/load project settings -> each window doesn't have to know about JSON for example. A `dict` is the most convenient type for this.
* for each type that will be saved to a project a pair of classes will be created, an Encoder and a Decoder
  * the first is responsible for creating a `dict` of the stored settings
  * the second is responsible for create the type from a `dict` of settings
* separate types for encoding/decoding rather than methods on each saveable type have advantages:
 * we can't have additional methods on `matplotlib` plots so separate classes is the only way there
 * we can write all project saving in python, even for C++ widgets
 * unit testing of the separate pieces will be more effective and less coupled together

I like the sequence diagram from below and maybe that should then follow this at the top of this section with then a paragraphs explaining each class in the sequence diagram?

Reply:
Agreed and Done :+1: